File: libraries/kafka-client-py/kafka_client_py/consumer/retry_policy.py
import random
from dataclasses import dataclass, field
from datetime import UTC, datetime, timedelta

from ..errors import PoisonError, TransientError


@dataclass(frozen=True, slots=True)
class RetryPolicy:
    max_in_process_attempts: int = 3
    max_retry_topic_attempts: int = 5

    initial_backoff: timedelta = timedelta(seconds=1)
    max_backoff: timedelta = timedelta(minutes=10)
    backoff_multiplier: float = 2.0
    jitter_ratio: float = 0.2

    retryable: tuple[type[BaseException], ...] = field(default_factory=tuple)

    def is_retryable(self, exception: BaseException) -> bool:
        if isinstance(exception, PoisonError):
            return False
        if isinstance(exception, TransientError):
            return True
        return isinstance(exception, self.retryable)

    def compute_backoff(self, retry_topic_attempt: int) -> timedelta:
        if retry_topic_attempt < 1:
            retry_topic_attempt = 1

        base_seconds = self.initial_backoff.total_seconds() * (
            self.backoff_multiplier ** (retry_topic_attempt - 1)
        )
        capped_seconds = min(base_seconds, self.max_backoff.total_seconds())

        if self.jitter_ratio > 0:
            jitter_spread = capped_seconds * self.jitter_ratio
            capped_seconds = capped_seconds + random.uniform(
                -jitter_spread,
                jitter_spread,
            )
            capped_seconds = max(capped_seconds, 0.0)

        return timedelta(seconds=capped_seconds)

    def compute_retry_at(
        self,
        retry_topic_attempt: int,
        *,
        now: datetime | None = None,
    ) -> datetime:
        base_timestamp = now or datetime.now(UTC)
        return base_timestamp + self.compute_backoff(retry_topic_attempt)

File: libraries/kafka-client-go/consumer/retry_policy.go
package consumer

import (
	"errors"
	"math"
	"math/rand/v2"
	"time"

	kafkaclient "github.com/power-finance/kafka-client-go"
)

type RetryPolicy struct {
	MaxInProcessAttempts  int
	MaxRetryTopicAttempts int

	InitialBackoff    time.Duration
	MaxBackoff        time.Duration
	BackoffMultiplier float64
	JitterRatio       float64

	Retryable     []error
	RetryableFunc func(error) bool
}

func DefaultRetryPolicy() RetryPolicy {
	return RetryPolicy{
		MaxInProcessAttempts:  3,
		MaxRetryTopicAttempts: 5,
		InitialBackoff:        time.Second,
		MaxBackoff:            10 * time.Minute,
		BackoffMultiplier:     2.0,
		JitterRatio:           0.2,
	}
}

func (p RetryPolicy) IsRetryable(err error) bool {
	if errors.Is(err, kafkaclient.ErrPoison) {
		return false
	} else if errors.Is(err, kafkaclient.ErrTransient) {
		return true
	}

	for _, target := range p.Retryable {
		if errors.Is(err, target) {
			return true
		}
	}

	return p.RetryableFunc != nil && p.RetryableFunc(err)
}

func (p RetryPolicy) ComputeBackoff(retryTopicAttempt int) time.Duration {
	if retryTopicAttempt < 1 {
		retryTopicAttempt = 1
	}

	baseSeconds := p.InitialBackoff.Seconds() * math.Pow(
		p.BackoffMultiplier,
		float64(retryTopicAttempt-1),
	)
	cappedSeconds := math.Min(baseSeconds, p.MaxBackoff.Seconds())

	if p.JitterRatio > 0 {
		jitterSpread := cappedSeconds * p.JitterRatio
		cappedSeconds += (rand.Float64()*2 - 1) * jitterSpread
		cappedSeconds = math.Max(cappedSeconds, 0.0)
	}

	return time.Duration(cappedSeconds * float64(time.Second))
}

func (p RetryPolicy) ComputeRetryAt(retryTopicAttempt int, now time.Time) time.Time {
	if now.IsZero() {
		now = time.Now().UTC()
	}
	return now.Add(p.ComputeBackoff(retryTopicAttempt))
}
